File: Searching/Ruby/breadth_first_search.rb
class Graph 
    def initialize 
        @graph = Hash.new { Array.new }
    end

    def add_edge(starting_vector, ending_vector)
        @graph[starting_vector] += [ending_vector]
    end

    def breadth_first_search(starting_vector)
        visited = Hash.new { false }

        queue = [starting_vector] + @graph[starting_vector]
        final_arr = []

        until queue.empty?
            current_vector = queue.shift

            if visited[current_vector] == false
                final_arr << current_vector
                visited[current_vector] = true
                queue += @graph[current_vector]
            end
        end

        puts final_arr
    end
end

graph = Graph.new

graph.add_edge(0, 1)
graph.add_edge(0, 2)
graph.add_edge(1, 2)
graph.add_edge(2, 0)
graph.add_edge(2, 3)
graph.add_edge(3, 3)

graph.breadth_first_search(2)

File: Searching/Python/breadth_first_search.py
from collections import defaultdict

class Graph:

    def __init__(self):
        self.graph = defaultdict(list)

    def addEdge(self, u, v):
        self.graph[u].append(v)

    def BFS(self, s):
        visited = [False] * (len(self.graph))

        queue = []

        queue.append(s)
        visited[s] = True

        while queue:
            s = queue.pop(0)
            print(s)

            for i in self.graph[s]:
                if visited[i] == False:
                    queue.append(i) 
                    visited[i] = True
    
graph = Graph() 

graph.addEdge(0, 1)
graph.addEdge(0, 2)
graph.addEdge(1, 2)
graph.addEdge(2, 0)
graph.addEdge(2, 3)
graph.addEdge(3, 3)

graph.BFS(2)